### PR TITLE
doc: Document how to analyze Bitcoin Core using Clang Static Analysis, clang-tidy and cppcheck

### PR DESCRIPTION
**Pull request by practicalswift**
Document how to analyze Bitcoin Core using Clang Static Analysis, `clang-tidy` and `cppcheck`.

As requested by Sjors (#14676) and others :)

### CODE REVIEW

File: doc/analysis.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+### Analyzing Bitcoin Core using Clang Static Analyzer
+
+#### Quickstart guide
+
+```
+$ git clone https://github.com/bitcoin/bitcoin
+$ cd bitcoin/
+$ ./autogen.sh
+$ CC=clang CXX=clang++ ./configure --with-incompatible-bdb --disable-ccache
```
Comment (jonatack) [inline doc/analysis.md:13]: It may be helpful to mention why `ccache` needs to be disabled.

**Review by jonatack — COMMENTED** [review]
Concept ACK. Good idea. I think each guide could begin with a sentence on why/what and not just how. Could squash to one commit.

File: doc/analysis.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+$ cd bitcoin/
+$ git clone https://github.com/danmar/cppcheck
+$ cd cppcheck/
+$ cmake -DFILESDIR=$(pwd)/ .
+$ make
+$ cd ..
+# Analyze source code files ...
+$ cppcheck/bin/cppcheck --language=c++ -D__cplusplus -DCLIENT_VERSION_BUILD \
+    -DCLIENT_VERSION_IS_RELEASE -DCLIENT_VERSION_MAJOR -DCLIENT_VERSION_MINOR \
+    -DCLIENT_VERSION_REVISION -DCOPYRIGHT_YEAR -DDEBUG
```
Comment (jonatack) [inline doc/analysis.md:52]: missing backslash at EOL

Comment (jonatack) [reply doc/analysis.md:52]: hehe, nice example
```bash
$ cppcheck/bin/cppcheck --language=c++ -D__cplusplus -DCLIENT_VERSION_BUILD \
> -DCLIENT_VERSION_IS_RELEASE -DCLIENT_VERSION_MAJOR -DCLIENT_VERSION_MINOR \
>  -DCLIENT_VERSION_REVISION -DCOPYRIGHT_YEAR -DDEBUG \
>  -DBOOST_FIXTURE_TEST_SUITE -I src/ -q \
> src/net_processing.cpp
src/net_processing.cpp:936:17: error: Same iterator is used with different containers 'mapOrphanTransactions' and 'itPrev.second'. [iterators1]
        itPrev->second.erase(it);
                ^
src/limitedmap.h:72:39: style: Same iterators expression are used for algorithm. [sameIteratorExpression]
        iterator itTarget = map.erase(itIn, itIn);
```

Comment (jonatack) [reply doc/analysis.md:52]: heh, nice example!
```bash
$ cppcheck/bin/cppcheck --language=c++ -D__cplusplus .../... src/net_processing.cpp
src/net_processing.cpp:936:17: error: Same iterator is used with different containers 'mapOrphanTransactions' and 'itPrev.second'. [iterators1]
        itPrev->second.erase(it);
                ^
src/limitedmap.h:72:39: style: Same iterators expression are used for algorithm. [sameIteratorExpression]
        iterator itTarget = map.erase(itIn, itIn);
```

File: doc/analysis.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+### Analyzing Bitcoin Core using Clang Static Analyzer
+
+#### Quickstart guide
+
+```
+$ git clone https://github.com/bitcoin/bitcoin
+$ cd bitcoin/
+$ ./autogen.sh
+$ CC=clang CXX=clang++ ./configure --with-incompatible-bdb --disable-ccache
+$ scan-build --use-cc=clang --use-c++=clang++ make
```
Comment (jonatack) [inline doc/analysis.md:14]: If this is OS-specific, I think it's necessary to state that. It seems to be, from looking at https://clang-analyzer.llvm.org/ and https://clang-analyzer.llvm.org/scan-build.html after seeing `scan-build: 'ccc-analyzer' does not exist at '/usr/local/bin/ccc-analyzer'` here and not finding a linux version.

File: doc/analysis.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+$ CC=clang CXX=clang++ ./configure --with-incompatible-bdb --disable-ccache
+$ git clone https://github.com/rizsotto/bear
+$ cd bear/
+$ cmake .
+$ make all
+$ chmod +x bear/bear
+$ cd ..
+$ bear/bear/bear -l $(pwd)/bear/libear/libear.so make
+# Analyze source code files ...
+$ clang-tidy src/test/crypto_tests.cpp
```
Comment (jonatack) [inline doc/analysis.md:34]: Is this also OS-specific?

**Comment by fanquake** [comment]
I think I'd rather see this kind of documentation added to the [devwiki](https://github.com/bitcoin-core/bitcoin-devwiki/wiki) (happy to help with that). That way it can be maintained / updated easier, without having to add even more documentation to this repo. There's also already other similar guides available, i.e I have one here: https://github.com/fanquake/core-review/blob/master/clang-tools.md.

**Comment by MarcoFalke** [comment]
There is also bitcoin-core/docs, if the devwiki or fanquake/core-review are insufficient places to put this.

In general I agree with @fanquake that documentation to compile with any imaginable compiler or checker or sanitizer or ... should probably be maintained outside of this repository. The project here is too big and moving too slow to keep every single piece of documentation up-to-date at all times. Also, we should be considerate of the precious review resource in this repository and use it on Bitcoin Core itself and not on integration of Bitcoin Core with meta developer tools.

**Comment by practicalswift** [comment]
@fanquake 

Feel free to take what you need from this PR and move to the appropriate place. Closing this PR :)

@marcofalke

> In general I agree with @fanquake that documentation to compile with any imaginable compiler or checker or sanitizer or ... should probably be maintained outside of this repository. The project here is too big and moving too slow to keep every single piece of documentation up-to-date at all times. Also, we should be considerate of the precious review resource in this repository and use it on Bitcoin Core itself and not on integration of Bitcoin Core with meta developer tools.

Personally I think we as a project have historically vastly underused the available tooling that is typically used in security critical projects to tame C++ (sanitizers, etc.). In other words I don't think we run the risk of using or documenting "too much tooling" (quite the opposite TBH!), but I see your point regarding moving the documentation to another repo :)

**Comment by MarcoFalke** [comment]
If one of the tools turns out to be useful for the greater project (for example be an optional but recommended tool for developers), I think it can surely be added to the documentation here.

**Comment by jonatack** [comment]
> Feel free to take what you need from this PR and move to the appropriate place. Closing this PR :)

Thanks for this info @practicalswift. Like @fanquake, I've been compiling info like this in a separate repository for my own reference and anyone else who may find it helpful.

**Comment by jonatack** [comment]
> similar guides available, i.e I have one here: [fanquake/core-review:clang-tools.md@`master`](https://github.com/fanquake/core-review/blob/master/clang-tools.md) [](/fanquake/core-review/blob/HEAD@{2020-05-19T12:55:20Z}/clang-tools.md).

Thanks @fanquake for the reminder to spend more time looking at the excellent info in core-review.

**Pull request closed without merging**